core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1001 @@
+   */
+  static class AddressTrackerMap extends ForwardingMap<EquivalentAddressGroup, AddressTracker> {
+    private final Map<EquivalentAddressGroup, AddressTracker> trackerMap;
+
+    AddressTrackerMap() {
+      trackerMap = new HashMap<>();
+    }
+
+    @Override
+    protected Map<EquivalentAddressGroup, AddressTracker> delegate() {
+      return trackerMap;
+    }
+
+    /** Adds a new tracker for the addresses that don't already have one. */
+    void putNewTrackers(OutlierDetectionLoadBalancerConfig config,
+        Collection<EquivalentAddressGroup> addresses) {
+      for (EquivalentAddressGroup address : addresses) {
+        if (!trackerMap.containsKey(address)) {
+          trackerMap.put(address, new AddressTracker(config));
+        }
+      }
+    }
+
+    /** Resets the call counters for all the trackers in the map. */
+    void resetCallCounters() {
+      for (AddressTracker tracker : trackerMap.values()) {
+        tracker.clearCallCounters();
+      }
+    }
+
+    /**
+     * When OD gets disabled we need to uneject any subchannels that may have been ejected and
+     * to reset the ejection time multiplier.
+     */
+    void cancelTracking() {
+      for (AddressTracker tracker : trackerMap.values()) {
+        if (tracker.subchannelsEjected()) {
+          tracker.unejectSubchannels();
+        }
+        tracker.resetEjectionTimeMultiplier();
+      }
+    }
+
+    /** Swaps the active and inactive counters for each tracker. */
+    void swapCounters() {
+      for (AddressTracker tracker : trackerMap.values()) {
+        tracker.swapCounters();
+      }
+    }
+
+    /**
+     * At the end of a timer run we need to decrement the ejection time multiplier for trackers
+     * that don't have ejected subchannels and uneject ones that have spent the maximum ejection
+     * time allowed.
+     */
+    void maybeUnejectOutliers(Long detectionTimerStartNanos) {
+      for (AddressTracker tracker : trackerMap.values()) {
+        if (!tracker.subchannelsEjected()) {
+          tracker.decrementEjectionTimeMultiplier();
+        }
+
+        if (tracker.subchannelsEjected() && tracker.maxEjectionTimeElapsed(
+            detectionTimerStartNanos)) {
+          tracker.unejectSubchannels();
+        }
+      }
+    }
+
+    /** Returns only the trackers that have the minimum configured volume to be considered. */
+    List<AddressTracker> trackersWithVolume(OutlierDetectionLoadBalancerConfig config) {
+      List<AddressTracker> trackersWithVolume = new ArrayList<>();
+      for (AddressTracker tracker : trackerMap.values()) {
+        if (tracker.inactiveVolume() >= config.successRateEjection.requestVolume) {
+          trackersWithVolume.add(tracker);
+        }
+      }
+      return trackersWithVolume;
+    }
+
+    /**
+     * How many percent of the addresses would have their subchannels ejected if we proceeded
+     * with the next ejection.
+     */
+    double nextEjectionPercentage() {
+      int totalAddresses = 0;
```

Comment:
It would be nice to explicitly return 0 when trackerMap.isEmpty() rather than rely on double handling divide by zero.